### ISSUE TITLE
Error when training mml parser on wikitables

### ISSUE THREAD

**Issue by YoPatapon** [allenai/allennlp#2053]
**Describe the bug**

I downloaded the DPD files from web and preprocessed the wikitables dataset with the provided preprocessing script into ".jsonl" files. Then I tried to run the program. However, I met such error with the default parameter setting in the jsonnet file.

> File "/home/v-dozhao/anaconda3/envs/py3.7torch/lib/python3.7/runpy.py", line 193, in _run_module_as_main
"main", mod_spec)
File "/home/v-dozhao/anaconda3/envs/py3.7torch/lib/python3.7/runpy.py", line 85, in _run_code
exec(code, run_globals)
File "/home/v-dozhao/Work/allennlp/allennlp/run.py", line 18, in 
main(prog="allennlp")
File "/home/v-dozhao/Work/allennlp/allennlp/commands/init.py", line 72, in main
args.func(args)
File "/home/v-dozhao/Work/allennlp/allennlp/commands/train.py", line 111, in train_model_from_args
args.force)
File "/home/v-dozhao/Work/allennlp/allennlp/commands/train.py", line 142, in train_model_from_file
return train_model(params, serialization_dir, file_friendly_logging, recover, force)
File "/home/v-dozhao/Work/allennlp/allennlp/commands/train.py", line 352, in train_model
metrics = trainer.train()
File "/home/v-dozhao/Work/allennlp/allennlp/training/trainer.py", line 755, in train
train_metrics = self._train_epoch(epoch)
File "/home/v-dozhao/Work/allennlp/allennlp/training/trainer.py", line 495, in _train_epoch
loss = self.batch_loss(batch, for_training=True)
File "/home/v-dozhao/Work/allennlp/allennlp/training/trainer.py", line 430, in batch_loss
output_dict = self.model(**batch)
File "/home/v-dozhao/anaconda3/envs/py3.7torch/lib/python3.7/site-packages/torch/nn/modules/module.py", line 477, in call
result = self.forward(*input, **kwargs)
File "/home/v-dozhao/Work/allennlp/allennlp/models/semantic_parsing/wikitables/wikitables_mml_semantic_parser.py", line 199, in forward
(target_action_sequences, target_mask))
File "/home/v-dozhao/Work/allennlp/allennlp/state_machines/trainers/maximum_marginal_likelihood.py", line 49, in decode
beam_search = ConstrainedBeamSearch(self._beam_size, targets, target_mask)
File "/home/v-dozhao/Work/allennlp/allennlp/state_machines/constrained_beam_search.py", line 58, in init
self._allowed_transitions = util.construct_prefix_tree(allowed_sequences, allowed_sequence_mask)
File "/home/v-dozhao/Work/allennlp/allennlp/state_machines/util.py", line 30, in construct_prefix_tree
assert targets.dim() == 3, "targets tensor needs to be batched!"
AttributeError: 'NoneType' object has no attribute 'dim'

I modified the batch size in the config file to 64 and I met a new error now:

> File "/home/v-dozhao/anaconda3/envs/py3.7torch/lib/python3.7/runpy.py", line 193, in _run_module_as_main
"main", mod_spec)
File "/home/v-dozhao/anaconda3/envs/py3.7torch/lib/python3.7/runpy.py", line 85, in _run_code
exec(code, run_globals)
File "/home/v-dozhao/Work/allennlp/allennlp/run.py", line 18, in 
main(prog="allennlp")
File "/home/v-dozhao/Work/allennlp/allennlp/commands/init.py", line 72, in main
args.func(args)
File "/home/v-dozhao/Work/allennlp/allennlp/commands/train.py", line 111, in train_model_from_args
args.force)
File "/home/v-dozhao/Work/allennlp/allennlp/commands/train.py", line 142, in train_model_from_file
return train_model(params, serialization_dir, file_friendly_logging, recover, force)
File "/home/v-dozhao/Work/allennlp/allennlp/commands/train.py", line 352, in train_model
metrics = trainer.train()
File "/home/v-dozhao/Work/allennlp/allennlp/training/trainer.py", line 755, in train
train_metrics = self._train_epoch(epoch)
File "/home/v-dozhao/Work/allennlp/allennlp/training/trainer.py", line 485, in _train_epoch
for batch in train_generator_tqdm:
File "/home/v-dozhao/anaconda3/envs/py3.7torch/lib/python3.7/site-packages/tqdm/_tqdm.py", line 979, in iter
for obj in iterable:
File "/home/v-dozhao/Work/allennlp/allennlp/data/iterators/data_iterator.py", line 143, in call
for batch in batches:
File "/home/v-dozhao/Work/allennlp/allennlp/data/iterators/basic_iterator.py", line 29, in _create_batches
batch = Batch(possibly_smaller_batches)
File "/home/v-dozhao/Work/allennlp/allennlp/data/dataset.py", line 33, in init
self._check_types()
File "/home/v-dozhao/Work/allennlp/allennlp/data/dataset.py", line 44, in _check_types
raise ConfigurationError("You cannot construct a Batch with non-homogeneous Instances.")
allennlp.common.checks.ConfigurationError: 'You cannot construct a Batch with non-homogeneous Instances.

**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.7
 - AllenNLP version: 0.7.1
 - PyTorch version: 0.4.1 post2

**Additional context**
I downloaded the DPD files from https://nlp.stanford.edu/software/sempre/wikitable/dpd/. And I modified the [code](https://github.com/allenai/allennlp/blob/43243acf4e91ba471923624bd48c9c9ec72332bf/allennlp/data/dataset_readers/semantic_parsing/wikitables/wikitables.py#L189-L202) that extract logical forms from dpd files because the data format is changed.

**Comment by YoPatapon** [comment]
I found the second error is caused by the missing of "target_action_sequences" field in some instances.

**Comment by matt-gardner** [comment]
Missing `target_action_sequences` is probably also the cause of the first problem, too.  What are you doing for instances that have no DPD output?

**Comment by YoPatapon** [comment]
I don't have any additional processing for instances without logical forms now. When I preprocessed the dpd files, I located the lines that starts with "derivation" in the file and try to extract the logic form using string operation. I didn't verify whether there is at least one logic form for in the instance.
Now I know not all instances have at least corresponding logical forms. What should we do with these instances? Should I modify the code to throw the instances away?

**Comment by YoPatapon** [comment]
OK, have you count how many instances are there that have no logical forms and how much does it affect the performance? By the way, does there exist a more convenient way to extract the logical form out now that the format is strange.

**Comment by matt-gardner** [comment]
There are somewhere around 7k examples that have DPD output for any particular training split, which is maybe 2/3 of the training data.  Yes, you could do better if you could make use of more data, but the question is how.  That's an open research problem.

For a "more convenient way to extract the logical form", nope, I don't have one.  I'd just pre-process the data with our script so that you can load it more efficiently the second time.

**Comment by YoPatapon** [comment]
I found many logical forms cannot be parsed by `expression = world.parse_logical_form(logical_form)` both when I preprocessed the data by the provided script and loaded data directly from ".examples" files. The logger outputs are like:

> 50it [00:24,  1.73it/s]2018-11-15 12:42:07,215 - DEBUG - allennlp.data.dataset_readers.semantic_parsing.wikitables.wikitables - Skipping logical form with inordinate number of "ors": ((reverse fb:row.row.director) (argmin (number 1) (number 1) (fb:row.row.details (fb:cell.cell.part (or (or (or (or fb:part.filmfare_best_director_puri_jagannadh fb:part.filmfare_best_music_director_sandeep_chowta) fb:part.nandi_award_for_best_art_director_s_ravinder) fb:part.nandi_award_for_best_director_ram_gopal_varma) fb:part.won_nandi_award_for_best_director_ram_gopal_varma))) (reverse (lambda x ((reverse fb:row.row.index) (fb:row.row.next (var x)))))))

Is this a normal situation?

**Comment by matt-gardner** [comment]
Yes, this is expected.  DPD produced a bunch of spurious logical forms, some of which take factorial time to parse with nltk (lots of nested "ors"), and we ignore them.  It _should_ be the case that all of the ignored logical forms are spurious (basically: lots of "ors", or numbers that don't show up in the question).  If you find a counter-example, let us know.

**Comment by YoPatapon** [comment]
I see. Thanks.

**Comment by YoPatapon** [comment]
@matt-gardner Hi Matt, can you provide the trained model that reports the best accuracy (which is 2-2.5% absolute below the accuracy mentioned in the paper) on the wikitables dataset. I trained with the default setting in the config file, however the model only got 34.7% accuracy on the corresponding dev set. 
And how many epochs does the model usually need to converge?

**Comment by matt-gardner** [comment]
That model was trained using old code, and won't load anymore with current code.  If you're using the unfiltered DPD file (which it looks like you are from other comments you've made), performance around 34-35% is expected (see [this paper](https://arxiv.org/abs/1803.04579)), and is 1-2% below the performance of the original implementation.  It was only with the filtered DPD file that the original model got higher accuracies.

However, we are experimenting with Chen Liang's language (from [this paper](https://arxiv.org/abs/1807.02322)).  Chen's MML model gets 39.9% accuracy on our split using his language and his initial search for logical forms.  @pdasigi used roughly the same data and language and got 39.3%.  This is _very_ late breaking (Pradeep just told me about this yesterday), and it's happening in our [research repo](https://github.com/allenai/allennlp-weak-supervision-research).

So, the bottom line is, you're seeing performance that's expected for the data and language that you're using.  We're working on better baselines, and Chen's language looks like it's substantially better than trying to use SEMPRE's lambda-DCS language (possibly because of mismatch in pre-processing / matching table entity names exactly for the SEMPRE executor, but also because the language is simpler).  We will merge this better baseline into the main repo eventually, but it will probably come after the NAACL deadline, as we're still iterating on things trying to get a submission ready on this stuff.  If you're really brave and willing to go without support, you can try the code in the research repository.  But we really can't answer questions about that code, and we might not have actually checked in everything you need to run stuff.

**Comment by YoPatapon** [comment]
@matt-gardner Thanks for your detailed explanation!
By the way, do we have a predicator for MML parser on wikitables now?

**Comment by matt-gardner** [comment]
Yes, there has been a [predictor for the wikitables parser](https://github.com/allenai/allennlp/blob/master/allennlp/predictors/wikitables_parser.py) in the repo for a long time.
